File: rot.py
def reverse_rot():
    with open("rot.in", 'r') as file: 
        alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ_."
        results = []
    
        for line in file:
            white_space = line.find(" ")
            if white_space == -1:
                continue  # Skip lines without a space
            
            n = int(line[:white_space])  # get the integer
            plain_text = line[white_space+1:].strip()  # get the entire string and strip newline
            reversed_s = plain_text[::-1]  # reverse the string
            
            if n == 0:
                break
            
            result = ''.join(alphabet[(alphabet.index(c) + n) % len(alphabet)] for c in reversed_s)
            results.append(result)
    
        print('\n'.join(results))

reverse_rot()

File: rot.pl
#!/usr/bin/perl
use strict;
use warnings;

sub reverse_rot {
    my $filename = 'rot.in';
    open(my $fh, '<', $filename) or die "Could not open file '$filename' $!";

    my $alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ_.";
    my $result = "";

    while (my $line = <$fh>) {
        chomp($line);  # Remove trailing newline
        my $white_space = index($line, ' ');  # Find the first space

        if ($white_space == -1) {
            next;  # Skip lines without a space
        }

        my $n = substr($line, 0, $white_space);  # Get the integer
        my $plain_text = substr($line, $white_space + 1);  # Get the text part
        my $reversed_s = reverse($plain_text);  # Reverse the string

        last if $n == 0;  # Exit the loop if n is 0

        foreach my $c (split //, $reversed_s) {
            my $x = index($alphabet, $c);
            $result .= substr($alphabet, ($x + $n) % length($alphabet), 1);
        }
        $result .= "\n";
    }

    close($fh);
    print $result;
}

reverse_rot();
